### PR TITLE
SDF strategy

### PR DESCRIPTION
**Pull request by dy**
SDF method to render fancy scatter. I reached `3e6` custom points.

**Changelog**

* Visible range of points is filtered with [snap-points-2d](https://github.com/gl-vis/snap-points-2d) and [binary-search-bounds](https://www.npmjs.com/package/binary-search-bounds), which solves [webgl memory overflow issue](https://github.com/plotly/plotly.js/issues/566), see #4 
* Characters are rendered with [SDF method](https://www.mapbox.com/blog/text-signed-distance-fields/) via [font-atlas-sdf](https://github.com/dfcreative/font-atlas-sdf), that extends data limits (up to 2e6) and removes heavy vectorization modules [text-cache](https://github.com/gl-vis/text-cache) and [vectorize-text](https://github.com/mikolalysenko/vectorize-text)
* Fixed bug of inverted text direction
* Position and other buffers squashed for optimized loading

**Limitations**

* Big number of big-sized points is now somewhat slower to render, hopefully it is rare case.
* Number of available colors is limited to 256
* Number of available characters is also limited, depending on point size

**TODO**
* [x] Render borders
* [x] Filter out rendering interval by boundaries
* [x] Make adjustable character texture size
* [x] Make antialias step absolute
* [x] Test font atlas character wrapping
* [x] Test various viewboxes
* [x] FIXMEs in code
* [x] Make sure no regression introduced, in particular hi-precision points, cross-browser stability, plotly.js tests
* [x] Add demo, elaborate readme

**Some observations**

* [ ] `scattergl` still renders up to `1e8` points, whereas this patch is only `1e6` points. That is due to size/borderWidth/char/color buffers, it is impossible to get rid of them. We can try to patch `gl-scatter2d` to take color variations though. Or we can think of an alternate scatter plot with multiple render passes for different data groups. Although for big data we need special data structures, like wavelets, see [ref](http://grail.cs.washington.edu/projects/wavelets/article/wavelet2.pdf).
* [ ] Interactions are slower than scattergl with 1e6 points due to heavier rendering. We can try to render texture with data and re-render on update only, not during the pick.
* [x] Point size is off a bit (`pixelRatio` was the reason)
* [x] Colors blending is different (turned out we need [special blending function](https://github.com/gl-vis/gl-scatter2d-fancy/blob/sdf/scatter-fancy.js#L144))
* [x] Texture error
* [ ] No point weight (need we?)

**Possible optimizations**
* Cache rendered font, grouped colors etc to reduce `update` calculations
* Re-render data only in `update` into texture, draw rendered texture in idle/pick mode

### CODE REVIEW

**Comment by dfcreative** [comment]
Some observations

* [ ] `scattergl` still renders up to 1e8 points, whereas here is only 1e6 points
* [ ] Interactions are noticeably slower than scattergl on 1e6 points
* [ ] Point size is off a bit
* [ ] Colors blending is different

**Comment by dy** [comment]
@monfera @etpinard could you please take a look/give it a spin?
`example/render` and `example/points` are basic test-cases, also https://github.com/plotly/plotly.js/issues/566 should be fixed.

Full review is going to be difficult...

**Comment by etpinard** [comment]
@dfcreative awesome! I'll take this for test :car: at some point today.

File: scatter-fancy.js [hunk 1]
```diff
@@ -4,116 +4,54 @@ module.exports = createFancyScatter2D
...
-  return BOUNDARIES[glyph] = {
-    coords:  coords,
-    normals: normals,
-    bounds:  bounds
-  }
-}
+var snapPoints = require('snap-points-2d')
+var atlas = require('font-atlas-sdf')
+var createTexture = require('gl-texture2d')
+var colorId = require('color-id')
```
Comment (etpinard) [inline scatter-fancy.js:12]: This [module](https://github.com/dfcreative/color-id/blob/master/index.js#L5) here contains some ES6 tokens that the (rather old) nw.js version we use in our image test docker container can't handle.

@dfcreative Would you mind switching those `const`s and `let`s to good `var`s?

Comment (etpinard) [reply scatter-fancy.js:12]: ... done in https://github.com/dfcreative/color-id/pull/1

**Comment by etpinard** [comment]
> Make sure no regression introduced, in particular hi-precision points, cross-browser stability, plotly.js tests

Ok. So I generated new baseline images using this branch. Go to

- https://github.com/plotly/plotly.js/compare/scattergl-sdf

for the details. I noticed a few things:

- Marker point borders are off. In plotly.js, marker point borders are specified by `marker.line.width`. For most mocks, the marker point borders appear a little thinner but in other cases e.g.

gifrecord_2017-02-10_153948

they appear much thicker!


- Marker point seem to be shifted slightly downward than on `master`. This issue is most noticeable in the [`gl2d_14`](https://github.com/plotly/plotly.js/compare/scattergl-sdf#diff-aca01895723b6b01df7932ca240a1d22) mock:

gifrecord_2017-02-10_153618

- [`gl2d_17`](https://github.com/plotly/plotly.js/blob/master/test/image/mocks/gl2d_17.json) is missing a few scatter points. Moreover, something messed up the line color

gifrecord_2017-02-10_154256

**Comment by etpinard** [comment]
Moreover, hover picking seems to not respect the marker size. This issue is most noticeable with [`gl2d_12`](https://github.com/plotly/plotly.js/blob/master/test/image/mocks/gl2d_12.json) mock:

On master:

gifrecord_2017-02-10_154920

On this branch:

gifrecord_2017-02-10_155017

**Comment by dy** [comment]
@etpinard thanks for the feedback, fixed everything except for marker displacement.
Turned out that marker is aligned a bit off inside of font itself:
image

Old _gl-scatter2d-fancy_ aligned the point vertically regardless of font baseline etc, based on glyph shape. Here we have a hard time doing that.

One way I see is replacing marker in [gl_markers](https://github.com/plotly/plotly.js/blob/master/src/constants/gl_markers.js) from `●` to `•` (← see how bolder bulled is lower) and adjusting bullet size.
The other way is trying to find a font with better vertical alignment for bullet.
Also I can manually adjust offset inside of the font (ad-hoc).

**Comment by dy** [comment]
@etpinard fixed everything. Turned out all marker glyphs used in plotly are shifted downwards within font, so we can safely use char offset.

The mentioned tests work alright except for point arrangement. How much is it of importance?

**Comment by etpinard** [comment]
Thanks @dfcreative ! I'll give this PR another look at some point today :tada:

**Comment by etpinard** [comment]
@dfcreative fantastic. I confirm that the hover picking issue of https://github.com/gl-vis/gl-scatter2d-fancy/pull/5#issuecomment-279062225 is fixed :white_check_mark: 

In addition, the new marker offset still generate a diff in the image test, but nothing to worry about at first glance :tada: 

Unfortunately, the `marker.line.width` computations are still off. See new diff [here](https://github.com/plotly/plotly.js/compare/scattergl-sdf#diff-40263d19b509de652cb33b21744665bb)

image


Moreover, when generating the new baseline one after the other, it some case the axes failed to be drawn e.g. for the [`gl2d_10`](https://github.com/plotly/plotly.js/commit/2822249202ea12482fb9e4578e8c151c8cee9996#diff-00f900c81fe62d8829de2b343a4651bf):

image


reminiscing of https://github.com/plotly/streambed/issues/6307

**Comment by dy** [comment]
@etpinard yes, sorry about colors, I've merged your PR to `color-id` without revision - turned out we can not check like that [here](https://github.com/dfcreative/color-id/commit/2735fb42f56e63df9706ac350603ce9a32f8815b#diff-168726dbe96b3ce427e7fedce31bb0bcL10).
As for missing axes - could you help me reproduce that? I am facing troubles getting it, I run
```js
const Plotly = require('plotly.js');

let list = [
	require('plotly.js/test/image/mocks/gl2d_12.json'),
	require('plotly.js/test/image/mocks/gl2d_14.json'),
	require('plotly.js/test/image/mocks/gl2d_17.json'),
	require('plotly.js/test/image/mocks/gl2d_10.json')
]

list.forEach(mock => {
	let div = document.body.appendChild(document.createElement('div'))
	Plotly.plot(div, mock.data, mock.layout).then((div) => {
		return Plotly.toImage(div).then(dataUrl => {
			let img = document.body.appendChild(document.createElement('img'))
			img.src = dataUrl
		})
	})
	.then(() => {
		Plotly.purge(div)
	})
})
```
and it's fine

**Comment by etpinard** [comment]
> yes, sorry about colors, I've merged your PR to color-id without revision - turned out we can not check like that here.

No worries! I pulled down the `color-id` latest changes and the `marker.line.width` are back to form! That is, at least in the browser, something is off in our image-test container though:

image

image

looks like the point ordering is off. But all is good in my browser window. Strange.

--------------------------

> could you help me reproduce that? I am facing troubles getting it, I run

Try:

```js
let list = [
	require('plotly.js/test/image/mocks/gl2d_10.json'),
	require('plotly.js/test/image/mocks/gl2d_12.json')
]
```

In the test dashboard, I get:

gifrecord_2017-02-15_144956

**Comment by etpinard** [comment]
@dfcreative before I forgot, can you check that @monfera 's http://codepen.io/monfera/full/VmYeQb/ still works off this branch. I think that [`gl2d_date_axes`](https://github.com/monfera/plotly.js/blob/master/test/image/baselines/gl2d_date_axes.png) does a good job checking for high-precision behavior, but you can't never be too sure :smile:

**Comment by etpinard** [comment]
... and for more info on our high-precision specs, see https://github.com/plotly/plotly.js/pull/1114

**Comment by dy** [comment]
@etpinard ok, fixed sorting order for small point sets and optimized @monfera's `gl2d_date_axes` case.
As for wrong scale of sequence of mocks — I am facing the same issue with old `gl-scatter2d-fancy` now, it seems to be an old symptom of some `gl-plot2d` problem. Trying to debug, but seems that is irrelevant to this PR.

UPD. increasing the delay [here](https://github.com/plotly/plotly.js/blob/master/src/snapshot/helpers.js#L18) up to `3000` fixes the issue. That is the only way to get around I see for now...

**Comment by dy** [comment]
@etpinard so after intense debugging I came up with [this PR](https://github.com/plotly/plotly.js/pull/1388). It fixes the problem of wrong axes in images.

**Comment by dy** [comment]
Actually mb it is better to create a separate repo, like _gl-scatter2d-sdf_? Changes are pretty significant, also loosing vectorizing logic completely is not too good. @etpinard

**Comment by etpinard** [comment]
> Actually mb it is better to create a separate repo, like gl-scatter2d-sdf?

I like this idea a lot. This will make comparing the two algorithms easier for future iterations.

> loosing vectorizing logic completely is not too good

For reference, could you elaborate a little more on the drawbacks of SDF? Is there anything else worth mentioning besides what's in _limitations_ section above?

> - Big number of big-sized points is now somewhat slower to render, hopefully it is rare case.
> - Number of available colors is limited to 256
> - Number of available characters is also limited, depending on point size

**Comment by dy** [comment]
@etpinard [done](https://github.com/gl-vis/gl-scatter2d-sdf).
Actually I could refactor API to overcome `1e6` limitation in future and speed up updating significantly. For now we require huge arrays of repeated data, but we could group it.
Two main drawbacks I see are
* it pre-renders canvas with font atlas, which is expected to be one-timer but potentially heavy operation (up to 100ms for some bad cases like 100+ different characters). That is avoidable by caching it.
* it might be somewhat longer to render for cases of multiple overlapping big-sized points.

**Comment by etpinard** [comment]
@dfcreative Amazing job. Here's how the baselines look like with your latest commits here and of https://github.com/plotly/plotly.js/pull/1388:

https://github.com/plotly/plotly.js/compare/scattergl-sdf

Everything looks in place to me. But I wouldn't mind having someone else look at the new baselines to double check :eyeglasses:

**Comment by etpinard** [comment]
On second glance, looks like the bubbles in the `gl2d_12` are sized slightly differently:

gifrecord_2017-02-17_135115

**Comment by dy** [comment]
@etpinard yes, that is due to adjusted [rendering](https://github.com/gl-vis/gl-scatter2d-sdf/blob/sdf/example/render.js), it looks better for the symbols.
image
vs
image
How much is that critical?

**Comment by etpinard** [comment]
@dfcreative I agree, the new rendering does look better for symbols :+1: 

But for bubbles charts, the size of each circle has an important meaning, so the new rendering in its current form could be considered a breaking change by some.

Would it be possible to some logic for `symbol: 'circle'` to make them look the way they do on `master`?

**Comment by dy** [comment]
@etpinard with https://github.com/gl-vis/gl-scatter2d-sdf/commit/707658c03647f97b1963f3e851696f44a5d8e538 I moved as close as I could to the initial fancy-scatter in size/border, but there are still tiny differences (1px-wise) due to method artefacts, it will be extremely difficult to remove them.

**Comment by etpinard** [comment]
@dfcreative 

https://github.com/gl-vis/gl-scatter2d-fancy/pull/5#issuecomment-280838250

off https://github.com/plotly/plotly.js/compare/gl-scatter2d-sdf

is looking good :tada: 

---

Can you publish the latest changes to [`gl-scatter2d-sdf`](https://www.npmjs.com/package/gl-scatter2d-sdf)? 

I'll take care of getting this in plotly.js master. Awesome work!

**Comment by dy** [comment]
Done https://www.npmjs.com/package/gl-scatter2d-sdf. Thank you!

**Pull request closed without merging**